File: dkim-txt-record-malformed/node/dkim-txt-check.js
/**
 * Detect a malformed DKIM TXT record and repair it via Cloudflare.
 * Safe to run on a schedule. Stays in dry run until DRY_RUN=false.
 */
import { pathToFileURL } from "node:url";

const TAG_RE = /(v|k|p)=([^;]*)/g;

export function validateDkimTxt(strings) {
  // Pure decision function. No network, no I/O.
  // strings is the list of character-strings the built-in dns module
  // returns for the TXT RRset at a selector, e.g.
  // ['v=DKIM1; k=rsa; p=MIIBIjq...', '...rest'].
  if (!strings || strings.length === 0) {
    return { valid: false, reason: "empty_key", keyBytes: null };
  }

  const joined = strings.join("");

  if (joined.includes('"') || joined.includes("\\")) {
    return { valid: false, reason: "embedded_quotes", keyBytes: null };
  }

  const tags = {};
  for (const match of joined.matchAll(TAG_RE)) {
    tags[match[1]] = match[2];
  }
  const pValue = (tags.p || "").trim();

  if (!pValue) {
    return { valid: false, reason: "empty_key", keyBytes: null };
  }

  if (/\s/.test(pValue)) {
    return { valid: false, reason: "embedded_quotes", keyBytes: null };
  }

  try {
    const keyBytes = Buffer.from(pValue, "base64");
    if (keyBytes.toString("base64").replace(/=+$/, "") !== pValue.replace(/=+$/, "")) {
      return { valid: false, reason: "not_base64", keyBytes: null };
    }
    return { valid: true, reason: "ok", keyBytes: keyBytes.length };
  } catch {
    return { valid: false, reason: "not_base64", keyBytes: null };
  }
}

export async function run() {
  // Imported lazily so the pure function above can be tested with no
  // network or crypto modules touched at all.
  const dns = await import("node:dns");
  const crypto = await import("node:crypto");

  const domain = process.env.DNS_DOMAIN;
  const selector = process.env.DKIM_SELECTOR || "selector1";
  const newRecordValue = process.env.DKIM_RECORD_VALUE;
  const ttl = Number(process.env.RECORD_TTL || 3600);
  const dryRun = (process.env.DRY_RUN || "true").toLowerCase() === "true";

  const zoneId = process.env.CLOUDFLARE_ZONE_ID;
  const apiToken = process.env.CLOUDFLARE_API_TOKEN;

  const name = `${selector}._domainkey.${domain}`;

  let rrsets = [];
  try {
    const records = await dns.promises.resolveTxt(name);
    rrsets = records;
  } catch (err) {
    if (err.code !== "ENOTFOUND" && err.code !== "ENODATA") throw err;
  }

  let outcome;
  if (rrsets.length > 1) {
    console.warn(`Selector ${selector} has ${rrsets.length} separate TXT RRsets at the same name (multiple_txt_records)`);
    outcome = { valid: false, reason: "multiple_txt_records", keyBytes: null };
  } else if (rrsets.length === 0) {
    outcome = { valid: false, reason: "empty_key", keyBytes: null };
  } else {
    outcome = validateDkimTxt(rrsets[0]);
  }

  console.log(`Selector ${selector} (${name}): valid=${outcome.valid} reason=${outcome.reason} keyBytes=${outcome.keyBytes}`);

  if (outcome.valid) {
    const joined = rrsets[0].join("");
    const match = /p=([^;]*)/.exec(joined);
    const pValue = match ? match[1].trim() : "";
    try {
      crypto.createPublicKey({
        key: Buffer.from(pValue, "base64"),
        format: "der",
        type: "spki",
      });
      console.log(`Key for ${name} loads as a valid public key.`);
    } catch (exc) {
      console.warn(`Key for ${name} decoded as base64 but did not load as a public key: ${exc.message}`);
    }
    return;
  }

  if (!newRecordValue) {
    console.log(`No DKIM_RECORD_VALUE set, skipping repair for ${selector}.`);
    return;
  }

  console.log(`${dryRun ? "Would" : "Will"} remove the broken record(s) and republish ${name}.`);
  if (dryRun) return;

  const lookupRes = await fetch(
    `https://api.cloudflare.com/client/v4/zones/${zoneId}/dns_records?type=TXT&name=${encodeURIComponent(name)}`,
    { headers: { Authorization: `Bearer ${apiToken}` } },
  );
  if (!lookupRes.ok) throw new Error(`Cloudflare lookup returned ${lookupRes.status}`);
  const lookupBody = await lookupRes.json();
  const existing = lookupBody.result || [];

  for (const record of existing) {
    const delRes = await fetch(
      `https://api.cloudflare.com/client/v4/zones/${zoneId}/dns_records/${record.id}`,
      { method: "DELETE", headers: { Authorization: `Bearer ${apiToken}` } },
    );
    if (!delRes.ok) throw new Error(`Cloudflare delete returned ${delRes.status}`);
    console.log(`Deleted broken TXT record ${record.id} at ${name}`);
  }

  const createRes = await fetch(
    `https://api.cloudflare.com/client/v4/zones/${zoneId}/dns_records`,
    {
      method: "POST",
      headers: { Authorization: `Bearer ${apiToken}`, "Content-Type": "application/json" },
      body: JSON.stringify({ type: "TXT", name, content: newRecordValue, ttl }),
    },
  );
  if (!createRes.ok) throw new Error(`Cloudflare create returned ${createRes.status}`);
  console.log(`Published corrected DKIM TXT record for ${name}`);
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: dkim-txt-record-malformed/python/dkim_txt_check.py
"""Detect a malformed DKIM TXT record and repair it via Cloudflare.
Safe to run on a schedule. Stays in dry run until DRY_RUN=false.
"""
import os
import re
import base64
import logging

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("dkim_txt_check")

TAG_RE = re.compile(r"(v|k|p)=([^;]*)")


def validate_dkim_txt(strings: list) -> dict:
    """Pure decision function. No network, no I/O.

    strings is the list of character-strings dnspython returns for the
    TXT RRset at a selector, e.g. ['v=DKIM1; k=rsa; p=MIIBIjq...', '...rest'].
    """
    if strings is None or len(strings) == 0:
        return {"valid": False, "reason": "empty_key", "key_bytes": None}

    joined = "".join(strings)

    if '"' in joined or "\\" in joined:
        return {"valid": False, "reason": "embedded_quotes", "key_bytes": None}

    tags = dict(TAG_RE.findall(joined))
    p_value = tags.get("p", "").strip()

    if not p_value:
        return {"valid": False, "reason": "empty_key", "key_bytes": None}

    if " " in p_value or "\n" in p_value or "\t" in p_value:
        return {"valid": False, "reason": "embedded_quotes", "key_bytes": None}

    try:
        key_bytes = base64.b64decode(p_value, validate=True)
    except Exception:
        return {"valid": False, "reason": "not_base64", "key_bytes": None}

    return {"valid": True, "reason": "ok", "key_bytes": len(key_bytes)}


def run():
    # Imported lazily so the pure function above can be tested with no
    # network or crypto libraries installed at all.
    import dns.resolver
    import requests
    from cryptography.hazmat.primitives.serialization import load_der_public_key

    domain = os.environ["DNS_DOMAIN"]
    selector = os.environ.get("DKIM_SELECTOR", "selector1")
    new_record_value = os.environ.get("DKIM_RECORD_VALUE")
    ttl = int(os.environ.get("RECORD_TTL", "3600"))
    dry_run = os.environ.get("DRY_RUN", "true").lower() == "true"

    zone_id = os.environ["CLOUDFLARE_ZONE_ID"]
    api_token = os.environ["CLOUDFLARE_API_TOKEN"]

    name = f"{selector}._domainkey.{domain}"

    try:
        answer = dns.resolver.resolve(name, "TXT")
        rrsets = []
        for rdata in answer:
            strings = [s.decode() if isinstance(s, bytes) else s for s in rdata.strings]
            rrsets.append(strings)
    except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer):
        rrsets = []

    if len(rrsets) > 1:
        log.warning("Selector %s has %d separate TXT RRsets at the same name (multiple_txt_records)", selector, len(rrsets))
        outcome = {"valid": False, "reason": "multiple_txt_records", "key_bytes": None}
    elif len(rrsets) == 0:
        outcome = {"valid": False, "reason": "empty_key", "key_bytes": None}
    else:
        outcome = validate_dkim_txt(rrsets[0])

    log.info("Selector %s (%s): valid=%s reason=%s key_bytes=%s", selector, name, outcome["valid"], outcome["reason"], outcome["key_bytes"])

    if outcome["valid"]:
        # Belt and suspenders: also confirm cryptography can load the key.
        p_value = dict(TAG_RE.findall("".join(rrsets[0]))).get("p", "").strip()
        try:
            load_der_public_key(base64.b64decode(p_value))
            log.info("Key for %s loads as a valid public key.", name)
        except Exception as exc:
            log.warning("Key for %s decoded as base64 but did not load as a public key: %s", name, exc)
        return

    if not new_record_value:
        log.info("No DKIM_RECORD_VALUE set, skipping repair for %s.", selector)
        return

    log.info("%s remove the broken record(s) and republish %s.", "Would" if dry_run else "Will", name)
    if dry_run:
        return

    lookup = requests.get(
        f"https://api.cloudflare.com/client/v4/zones/{zone_id}/dns_records",
        headers={"Authorization": f"Bearer {api_token}"},
        params={"type": "TXT", "name": name},
        timeout=30,
    )
    lookup.raise_for_status()
    existing = lookup.json().get("result", [])

    for record in existing:
        del_resp = requests.delete(
            f"https://api.cloudflare.com/client/v4/zones/{zone_id}/dns_records/{record['id']}",
            headers={"Authorization": f"Bearer {api_token}"},
            timeout=30,
        )
        del_resp.raise_for_status()
        log.info("Deleted broken TXT record %s at %s", record["id"], name)

    create_resp = requests.post(
        f"https://api.cloudflare.com/client/v4/zones/{zone_id}/dns_records",
        headers={"Authorization": f"Bearer {api_token}", "Content-Type": "application/json"},
        json={"type": "TXT", "name": name, "content": new_record_value, "ttl": ttl},
        timeout=30,
    )
    create_resp.raise_for_status()
    log.info("Published corrected DKIM TXT record for %s", name)


if __name__ == "__main__":
    run()
